Mathlib/Algebra/Module/FinitePresentation.lean

```diff
@@ -0,0 +1,285 @@
+  have hσ₂ : l ∘ b ∘ σ = Subtype.val := by ext i; exact congr_arg Subtype.val (hσ i)
+  refine ⟨(Set.finite_range (l ∘ b)).toFinset,
+    by simpa [Set.range_comp, LinearMap.range_eq_top], ?_⟩
+  let f : M →ₗ[R] (Set.finite_range (l ∘ b)).toFinset →₀ R :=
+    Finsupp.lmapDomain _ _ π ∘ₗ b.repr.toLinearMap
+  convert hl'.map f
+  ext x; simp only [LinearMap.mem_ker, Submodule.mem_map]
+  constructor
+  · intro hx
+    refine ⟨b.repr.symm (x.mapDomain σ), ?_, ?_⟩
+    · simp [Finsupp.apply_total, hσ₂, hx]
+    · simp only [f, LinearMap.comp_apply, b.repr.apply_symm_apply,
+        LinearEquiv.coe_toLinearMap, Finsupp.lmapDomain_apply]
+      rw [← Finsupp.mapDomain_comp, hσ₁, Finsupp.mapDomain_id]
+  · rintro ⟨y, hy, rfl⟩
+    simp [f, hπ, ← Finsupp.apply_total, hy]
+
+-- Ideally this should be an instance but it makes mathlib much slower.
+variable (R M) in
+lemma Module.finitePresentation_of_free [Module.Free R M] [Module.Finite R M] :
+    Module.FinitePresentation R M :=
+  Module.finitePresentation_of_free_of_surjective LinearMap.id (⟨·, rfl⟩)
+    (by simpa using Submodule.fg_bot)
+
+variable {ι} [_root_.Finite ι]
+
+instance : Module.FinitePresentation R R := Module.finitePresentation_of_free _ _
+instance : Module.FinitePresentation R (ι →₀ R) := Module.finitePresentation_of_free _ _
+instance : Module.FinitePresentation R (ι → R) := Module.finitePresentation_of_free _ _
+
+lemma Module.finitePresentation_of_surjective [h : Module.FinitePresentation R M] (l : M →ₗ[R] N)
+    (hl : Function.Surjective l) (hl' : Submodule.FG (LinearMap.ker l)) :
+    Module.FinitePresentation R N := by
+  classical
+  obtain ⟨s, hs, hs'⟩ := h
+  obtain ⟨t, ht⟩ := hl'
+  have H : Function.Surjective (Finsupp.total s M R Subtype.val) :=
+    LinearMap.range_eq_top.mp (by rw [Finsupp.range_total, Subtype.range_val, ← hs]; rfl)
+  apply Module.finitePresentation_of_free_of_surjective (l ∘ₗ Finsupp.total s M R Subtype.val)
+    (hl.comp H)
+  choose σ hσ using (show _ from H)
+  have : Finsupp.total s M R Subtype.val '' (σ '' t) = t := by
+    simp only [Set.image_image, hσ, Set.image_id']
+  rw [LinearMap.ker_comp, ← ht, ← this, ← Submodule.map_span, Submodule.comap_map_eq,
+    ← Finset.coe_image]
+  exact Submodule.FG.sup ⟨_, rfl⟩ hs'
+
+lemma Module.FinitePresentation.fg_ker [Module.Finite R M]
+    [h : Module.FinitePresentation R N] (l : M →ₗ[R] N) (hl : Function.Surjective l) :
+    Submodule.FG (LinearMap.ker l) := by
+  classical
+  obtain ⟨s, hs, hs'⟩ := h
+  have H : Function.Surjective (Finsupp.total s N R Subtype.val) :=
+    LinearMap.range_eq_top.mp (by rw [Finsupp.range_total, Subtype.range_val, ← hs]; rfl)
+  obtain ⟨f, hf⟩ : ∃ f : (s →₀ R) →ₗ[R] M, l ∘ₗ f = (Finsupp.total s N R Subtype.val) := by
+    choose f hf using show _ from hl
+    exact ⟨Finsupp.total s M R (fun i ↦ f i), by ext; simp [hf]⟩
+  apply Submodule.fg_of_fg_map_of_fg_inf_ker f.range.mkQ
+  · convert Module.Finite.out
```

Comment:
`convert` followed by a bunch of tactics is sometimes a style which is hard to maintain. Is there a cheap way of avoiding the `convert`? If not then don't worry.